### PR TITLE
remove .data from test/distributed/

### PR DESCRIPTION
**Pull request by izdeby**
`.data` calls are unsafe and should not be used.

### CODE REVIEW

File: test/distributed/test_c10d.py [hunk 1]
```diff
@@ -1980,7 +1980,6 @@ def step_model(model, input, target):
 
         def update_parameters(model):
             for param in model.parameters():
-                param.data -= param.grad
```
Comment (albanD) [inline test/distributed/test_c10d.py]: This one is actually modifying the value of `param` initially.
I think you want:
```
with torch.no_grad():
  param -= param.grad
```

File: test/distributed/test_data_parallel.py [hunk 2]
```diff
@@ -132,8 +132,8 @@ def forward(self, var1, var2, float1, var3=None):
         gvar2_exp = var2.grad.clone()
 
         def local_test(out):
-            var1.grad.data.fill_(0.0)
-            var2.grad.data.fill_(0.0)
+            var1.grad.fill_(0.0)
```
Comment (albanD) [inline test/distributed/test_data_parallel.py]: Is it fine to do this update in a differentiable manner? Or do we want a `with torch.no_grad():` ? 
Same below?

File: test/distributed/test_data_parallel.py [hunk 3]
```diff
@@ -531,10 +531,10 @@ def _test_scatter(self, tensor):
         self.assertEqual(result[0].get_device(), 0)
         self.assertEqual(result[1], x[2:])
         self.assertEqual(result[1].get_device(), 1)
-        grad = result[0].data.clone().fill_(2)
+        grad = result[0].clone().fill_(2)
```
Comment (albanD) [inline test/distributed/test_data_parallel.py]: I think this ones needs a `.detach()` no? You don't want grad to share the history of result?

File: test/distributed/test_distributed.py [hunk 4]
```diff
@@ -1731,7 +1731,6 @@ def test_all_gather_multigpu(self):
     def _model_step(self, model):
         for param in model.parameters():
             if param.grad is not None:
-                param.data += param.grad
```
Comment (albanD) [inline test/distributed/test_distributed.py]: Same as above about updating the value of `param`.

**Review by albanD — CHANGES_REQUESTED** [review]
Mostly good, just some unsure points.

**Review by albanD — APPROVED** [review]
LGTM, thanks !
Just small lint failures to fix.
